kura/org.eclipse.kura.wire.component.provider/src/main/java/org/eclipse/kura/internal/wire/subscriber/CloudSubscriber.java

```diff
@@ -54,32 +59,81 @@
  * Component, the user can only avail to wrap every Wire Record in the default
  * Google Protobuf Payload.
  */
-public final class CloudSubscriber implements WireEmitter, ConfigurableComponent, DataServiceListener {
+public final class CloudSubscriber implements WireEmitter, ConfigurableComponent, CloudClientListener {
+
+    /**
+     * Inner class defined to track the CloudServices as they get added, modified or removed.
+     * Specific methods can refresh the cloudService definition and setup again the Cloud Client.
+     *
+     */
+    private final class CloudSubscriberServiceTrackerCustomizer
+            implements ServiceTrackerCustomizer<CloudService, CloudService> {
+
+        @Override
+        public CloudService addingService(ServiceReference<CloudService> reference) {
+            CloudSubscriber.this.cloudService = CloudSubscriber.this.componentContext.getBundleContext()
+                    .getService(reference);
+            try {
+                // recreate the Cloud Client
+                setupCloudClient();
+                subscribeTopic();
+            } catch (KuraException e) {
+                logger.error(wireMessages.cloudClientSetupProblem() + ThrowableUtil.stackTraceAsString(e));
+            }
+            return CloudSubscriber.this.cloudService;
+        }
+
+        @Override
+        public void modifiedService(ServiceReference<CloudService> reference, CloudService service) {
+            CloudSubscriber.this.cloudService = CloudSubscriber.this.componentContext.getBundleContext()
+                    .getService(reference);
+            try {
+                // recreate the Cloud Client
+                setupCloudClient();
+                subscribeTopic();
+            } catch (KuraException e) {
+                logger.error(wireMessages.cloudClientSetupProblem() + ThrowableUtil.stackTraceAsString(e));
+            }
+        }
+
+        @Override
+        public void removedService(ServiceReference<CloudService> reference, CloudService service) {
+            CloudSubscriber.this.cloudService = null;
+        }
+    }
 
     /** The Logger instance. */
-    private static final Logger s_logger = LoggerFactory.getLogger(CloudSubscriber.class);
+    private static final Logger logger = LoggerFactory.getLogger(CloudSubscriber.class);
 
     /** Localization Resource */
-    private static final WireMessages s_message = LocalizationAdapter.adapt(WireMessages.class);
+    private static final WireMessages wireMessages = LocalizationAdapter.adapt(WireMessages.class);
+
+    private ServiceTracker<CloudService, CloudService> cloudServiceTracker;
 
     /** The cloud service. */
     private volatile CloudService cloudService;
 
-    /** The data service. */
-    private volatile DataService dataService;
+    private CloudClient cloudClient;
+
+    private final ServiceTrackerCustomizer<CloudService, CloudService> customizer = new CloudSubscriberServiceTrackerCustomizer();
```

Comment:
the instance of the tracker must not be created eagerly because, if this bundle is not active (in case the dependencies are not satisfied), the instance of the tracker will still be created.